agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
+                    "Fails to parse {} to DeviceUsage ",
+                    device_usage_string.unwrap()
+                ),
+            )
+        })?;
+        match get_device_usage_state(&device_usage, node_name) {
+            DeviceUsageStatus::Free => {
+                let new_device_usage =
+                    DeviceUsage::create(&device_usage_kind, node_name).map_err(|e| {
+                        Status::new(
+                            Code::Unknown,
+                            format!("Fails to create DeviceUsage - {}", e),
+                        )
+                    })?;
+                instance
+                    .device_usage
+                    .insert(device_usage_id.to_string(), new_device_usage.to_string());
+
+                if let Err(e) = kube_interface
+                    .update_instance(&instance, instance_name, instance_namespace)
+                    .await
+                {
+                    if x == (MAX_INSTANCE_UPDATE_TRIES - 1) {
+                        trace!("try_update_instance_device_usage - update_instance returned error [{}] after max tries ... returning error", e);
+                        return Err(Status::new(Code::Unknown, "Could not update Instance"));
+                    }
+                    random_delay().await;
+                } else {
+                    return Ok(());
                 }
-                random_delay().await;
-            } else {
-                return Ok(());
             }
-        } else {
-            // Instance slot already reserved for this node
-            return Ok(());
-        }
+            DeviceUsageStatus::ReservedByConfiguration(_) => {
+                if matches!(device_usage_kind, DeviceUsageKind::Configuration(_)) {
+                    return Ok(());
+                } else {
+                    return Err(Status::new(
+                        Code::Unknown,
+                        "Requested device already in use",
+                    ));
+                }
+            }
+            DeviceUsageStatus::ReservedByInstance => {
+                if matches!(device_usage_kind, DeviceUsageKind::Instance) {
+                    return Ok(());
+                } else {
+                    return Err(Status::new(
+                        Code::Unknown,
+                        "Requested device already in use",
+                    ));
+                }
+            }
+            DeviceUsageStatus::ReservedByOtherNode => {
+                trace!("try_update_instance_device_usage - request for device slot {} previously claimed by a diff node {} than this one {} ... indicates the device on THIS node must be marked unhealthy, invoking ListAndWatch ... returning failure, next scheduling should succeed!",
+                    device_usage_id, device_usage.get_usage_name(), node_name);
+                return Err(Status::new(
+                    Code::Unknown,
+                    "Requested device already in use",
```

Comment:
nit: add some context ... maybe "Requested device already in use by another node"

Reply:
update error messages